decouple logsstream from transport layer

wdbaruni opened the pull request:
LogStream is tightly coupled to libp2p to stream logs from the compute node to the requester, which makes supporting NATS more challenging https://github.com/bacalhau-project/expanso-planning/issues/413

This PR decouples LogStream and makes fetching logs from the executor and compute node agnostic to how they are published to the requester.

The way this is achieved by utilizing channels for async vending of logs. The compute nodes return a channel for `AsyncResult[models.ExecutionLog]`, which is consumed by the transport layer (`bprotocol` in current implementation) and sent to the requester. A similar implementation is also introduced from the client to the requester node where instead of exposing websockets to the client user, the new SDK just returns a channel of `AsyncResult[models.ExecutionLog]`

This PR also migrates `logs` endpoint from the old APIs to the new version, implements logs SDK, and introduces helper methods to expose and consume additional websockets in the future.

wdbaruni commented:
@frrist as our golang expert, I would appreciate your review on the use of channels in this PR

frrist reviewed (commented):
Left a couple early review comments.

I have also tried this locally. The command:
```bash
bacalhau --api-host=0.0.0.0 docker run --follow ubuntu:latest -- "bash" "-c" 'for i in {1..60}; do echo "Iteration $i"; sleep 1; done'
```
works as expected. 
However both `job logs` and `logs` just return exit code zero with no output.

EDIT: I had assumed the `logs` and `job logs` commands would follow the logs, but follow defaults to `false`. I further assumed I would at least see the logs emitted since the job had started, but we use `time.Now()` if the request doesn't contain `WithHistory` - and there isn't a way to specify that on `job logs` so the value is also `false`. This means the default behavior of the `job logs` command is unfriendly to users - it won't follow the logs, and it will show you the logs that were emitted around the time the command was run, which is racy at best - almost always no output.

The action I would propose taking here is:
- ensure `--history` can be provided to the `job logs` command (currently there is no flag)
- set the default value of `history` to `true`

This will ensure users see at least the logs since the beginning of the execution rather than noting, or a single line if they time it right.

frrist commented on pkg/transport/bprotocol/constants.go:
```diff
@@ -6,7 +6,7 @@ const (
 	BidRejectedProtocolID = "/bacalhau/compute/bid_rejected/1.0.0"
 	CancelProtocolID      = "/bacalhau/compute/cancel/1.0.0"
-	ExecutionLogsID       = "/bacalhau/compute/executionlogs/1.0.0"
+	ExecutionLogsID       = "/bacalhau/compute/execution_logs/1.0.0"
```
I think this is going to break compatibility with all previous version of this, any reason to change this?

wdbaruni replied:
yeah we have a new protocol for log vending between the compute node and the requester. Meaning we now need both compute and requester nodes to be on the same version for logging to work. We can utilize `NewMinVersionNodeRanker` to make sure requester won't communicate with lower version compute nodes, but I've checked our graphana metrics for how often users call our current logging API, and I barely saw 4 calls the past month. 

This should be a concern since it is not an actively used API, and just a client update is needed

frrist replied:
> yeah we have a new protocol for log vending between the compute node and the requester. 

Could we just bump the protocol version then?

frrist commented on pkg/publicapi/endpoint/orchestrator/job.go:
```diff
@@ -344,3 +347,73 @@ func (e *Endpoint) jobResults(c echo.Context) error {
+		return fmt.Errorf("failed to open log stream for job %s: %w", jobID, err)
+	}
+
+	for {
```
purely a style nit, but slight preference for:
```suggestion
	for logMsg := range logstreamCh {
		if err = ws.WriteJSON(logMsg); err != nil {
		        return err
		}
	}
```

wdbaruni replied:
Fixed in https://github.com/bacalhau-project/bacalhau/pull/3354/commits/80d1eef8857a74bd9ceb967d9958041bbcd42308

frrist commented on pkg/logger/dataframe.go:
```diff
@@ -28,38 +28,36 @@ const (
 	df.Data = make([]byte, df.Size)
 
 	n, err = reader.Read(df.Data)
 	if err != nil {
-		return DataFrame{}, err
+		return nil, err
```
This seems to be returning `EOF` before the container is done, related to my comment about `job logs` not working for me locally.

frrist commented on pkg/transport/bprotocol/compute_proxy.go:
```diff
@@ -146,5 +149,60 @@ func proxyRequest[Request any, Response any](
+			if err != nil {
+				_ = stream.Reset()
+				if err != io.EOF {
+					response.Err = fmt.Errorf("%s: failed to decode response from peer %s: %w", reflect.TypeOf(request), destPeerID, err)
```
`%T` should work instead of reflect.

wdbaruni replied:
Fixed in https://github.com/bacalhau-project/bacalhau/pull/3354/commits/80d1eef8857a74bd9ceb967d9958041bbcd42308

frrist commented on pkg/lib/concurrency/async_result.go:
```diff
@@ -0,0 +1,51 @@
+		Error string `json:"error,omitempty"`
+		*Alias
+	}{
+		Error: errorToString(ar.Err),
```
could also do:
```suggestion
                Error: fmt.Sprintf("%s",ar.Err)
```
instead of a new method.

wdbaruni replied:
not efficient, specially that this is going to be called each time a log record is serialized and deserialized when sent over the wire. A new method might be justified here

rossjones commented on pkg/compute/logstream/live_streamer.go:
```diff
@@ -0,0 +1,78 @@
+
+type LiveStreamerParams struct {
+	Reader io.Reader
+	Buffer int
```
This is set but I don't think it's used?

wdbaruni replied:
good catch. fixed in  https://github.com/bacalhau-project/bacalhau/pull/3354/commits/81fe8c5645703d466ae06b2ed828aac4e5fd2bf4

rossjones commented on pkg/nats/proxy/compute_proxy.go:
```diff
@@ -50,10 +52,10 @@ func (p *ComputeProxy) CancelExecution(
-		ctx, p.conn, request.TargetPeerID, ExecutionLogs, request)
+func (p *ComputeProxy) ExecutionLogs(ctx context.Context, request compute.ExecutionLogsRequest) (
+	<-chan *concurrency.AsyncResult[models.ExecutionLog], error) {
+	//TODO implement me
```
Is this coming in a follow up PR or a later commit to this one?

wdbaruni replied:
yeah in a follow up PR. I have an open issue for that https://github.com/bacalhau-project/expanso-planning/issues/413

wdbaruni commented on pkg/compute/logstream/completed_streamer.go:
```diff
@@ -0,0 +1,61 @@
+	}
+}
+
+func (s *CompletedStreamer) Stream(ctx context.Context) chan *concurrency.AsyncResult[models.ExecutionLog] {
+	ch := make(chan *concurrency.AsyncResult[models.ExecutionLog])
+	go func() {
+		defer close(ch)
+		if s.execution.RunOutput != nil {
+			s.process(ch, s.execution.RunOutput.STDOUT, models.ExecutionLogTypeSTDOUT)
+			s.process(ch, s.execution.RunOutput.STDERR, models.ExecutionLogTypeSTDERR)
+		}
+	}()
+	return ch
+}
+
+func (s *CompletedStreamer) process(ch chan *concurrency.AsyncResult[models.ExecutionLog], output string, typ models.ExecutionLogType) {
+	scanner := bufio.NewScanner(strings.NewReader(output))
+	for scanner.Scan() {
+		asyncResult := concurrency.AsyncResult[models.ExecutionLog]{
+			Value: models.ExecutionLog{
+				Type: typ,
+				Line: scanner.Text() + "\n",
+			},
+		}
+		ch <- &asyncResult
+	}
+	if scanner.Err() != nil {
+		asyncResult := concurrency.AsyncResult[models.ExecutionLog]{
+			Err: fmt.Errorf("failed to read output: %w", scanner.Err()),
+		}
+		ch <- &asyncResult
+	}
+}
```
Good point. That would allow `process` to exit if the channel was full and the context is done, right?

wdbaruni replied:
Done in https://github.com/bacalhau-project/bacalhau/pull/3354/commits/80d1eef8857a74bd9ceb967d9958041bbcd42308

frrist replied:
> That would allow process to exit if the channel was full and the context is done, right?

If the channel is full it will block (thus my comments about buffers) if ctx is done it exits.

wdbaruni commented on pkg/config/configenv/dev.go:
```diff
@@ -147,6 +147,9 @@ var DevelopmentComputeConfig = types.ComputeConfig{
 		Frequency: types.Duration(1 * time.Hour),
 	},
+	LogStreamConfig: types.LogStreamConfig{
+		ChannelBufferSize: 10,
```
this is number of log lines and not bytes. How many `models.ExecutionLog` can fit in the channel before we block if we have a slow channel consumer

wdbaruni commented on cmd/util/logging.go:
```diff
@@ -31,142 +33,53 @@ func Logs(cmd *cobra.Command, jobID string, follow, history bool) error {
+func readLogoutput(ctx context.Context, logsChannel <-chan *concurrency.AsyncResult[models.ExecutionLog]) error {
 	interrupt := make(chan os.Signal, 1)
 	signal.Notify(interrupt, os.Interrupt)
 
```
Done in https://github.com/bacalhau-project/bacalhau/pull/3354/commits/50012081cddce56e38eb7a30095ff64ff6a9790c

Though I am sure we can cleanup interruption handling from other commands as well

rossjones reviewed (commented):
LGTM. As per Frrst's comments, I believe the intent is that `logs` will show the history and not follow, whilst -f would show the history AND follow so defaults should reflect that.

Haven't approved because of forrest's comments.

wdbaruni commented on cmd/util/logging.go:
```diff
@@ -31,142 +33,53 @@ func Logs(cmd *cobra.Command, jobID string, follow, history bool) error {
-			return nil
 		case <-ctx.Done():
-			exiting = true
 			return nil
```
Fixed in https://github.com/bacalhau-project/bacalhau/pull/3354/commits/80d1eef8857a74bd9ceb967d9958041bbcd42308

wdbaruni commented:
I agree with showing the history to be enabled by default, though I hate making a boolean value true by default as we will end up with inconsistent cli and api behaviours. I'll change `history` to `tail`, and make it false by default. Does that make sense?

wdbaruni commented:
Switched to using `--tail` and also added `--execution-id` to allow users to fetch logs for a specific execution instead of letting the requester chose one
https://github.com/bacalhau-project/bacalhau/pull/3354/commits/c9d07be61cdc23e289dbc1630b74c638284af275

frrist commented on cmd/cli/job/logs.go:
```diff
@@ -36,16 +40,31 @@ func NewLogCmd() *cobra.Command {
 		},
 	}
 
+	logsCmd.PersistentFlags().StringVarP(
```
❤️
